### PR TITLE
Solve System.InvalidOperationException: Collection was modified

### PR DESCRIPTION
**Pull request by Neurosploit**
```
The following error occurred while creating wallet

[2:55] 
crit: Stratis.Bitcoin.Consensus[0]
      Consensus loop unhandled exception (Tip:118471)
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at Stratis.Bitcoin.Wallet.WalletManager.UpdateLastBlockSyncedHeight(ChainedBlock chainedBlock) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Wallet\WalletManager.cs:line 881
   at Stratis.Bitcoin.Wallet.WalletManager.ProcessBlock(Block block, ChainedBlock chainedBlock) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Wallet\WalletManager.cs:line 629
   at Stratis.Bitcoin.Wallet.WalletSyncManager.ProcessBlock(Block block) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Wallet\WalletSyncManager.cs:line 134
   at Stratis.Bitcoin.Wallet.Notifications.BlockObserver.OnNextCore(Block block) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Wallet\Notifications\BlockObserver.cs:line 24
   at System.Reactive.SafeObserver`1.OnNext(TSource value)
   at System.Reactive.Observer`1.OnNext(T value)
   at System.Reactive.SynchronizedObserver`1.OnNextCore(T value)
   at Stratis.Bitcoin.Signaler`1.Broadcast(T item) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Signaler.cs:line 46
   at Stratis.Bitcoin.Consensus.ConsensusFeature.RunLoop() in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Consensus\ConsensusFeature.cs:line 149

Unhandled Exception: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at Stratis.Bitcoin.Wallet.WalletManager.UpdateLastBlockSyncedHeight(ChainedBlock chainedBlock) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Wallet\WalletManager.cs:line 881
   at Stratis.Bitcoin.Wallet.WalletManager.ProcessBlock(Block block, ChainedBlock chainedBlock) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Wallet\WalletManager.cs:line 629
   at Stratis.Bitcoin.Wallet.WalletSyncManager.ProcessBlock(Block block) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Wallet\WalletSyncManager.cs:line 134
   at Stratis.Bitcoin.Wallet.Notifications.BlockObserver.OnNextCore(Block block) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Wallet\Notifications\BlockObserver.cs:line 24
   at System.Reactive.SafeObserver`1.OnNext(TSource value)
   at System.Reactive.Observer`1.OnNext(T value)
   at System.Reactive.SynchronizedObserver`1.OnNextCore(T value)
   at Stratis.Bitcoin.Signaler`1.Broadcast(T item) in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Signaler.cs:line 46
   at Stratis.Bitcoin.Consensus.ConsensusFeature.RunLoop() in C:\Users\stratis dev\Documents\StratisBitcoinFullNode\Stratis.Bitcoin\Consensus\ConsensusFeature.cs:line 168
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
```

Made a test for it to prove it is solved. We have to consider using ConcurrentBag or a different thread safe collection instead of List<T> in the future because List<T> is not thread safe and do not handle concurrency well. This will lead to issues with the signal/observer pattern.

### CODE REVIEW

**Comment by dangershony** [comment]
Ah we where just yesterday discussing threading safety on the wallet, I thought it would be very rare (apparently not).

There actually other places we need to protect against concurrency.
Each account has internal lists of addresses and each address has an internal list of transactions.
The places we might hit concurrency issues are: 
- Build transactions
- Send Transactions
- Periodic SaveWallet
- Add wallet
- History
- Balance

Basically anything that hits the wallet outside of the signaler. 
We might want to use locks instead of concurrent lists.

**Comment by Neurosploit** [comment]
I'll test it with a lock tomorrow.

File: Stratis.Bitcoin/Wallet/WalletManager.cs [hunk 1]
```diff
@@ -897,10 +898,9 @@ public void SaveToFile(Wallet wallet)
         /// <inheritdoc />
         public void UpdateLastBlockSyncedHeight(ChainedBlock chainedBlock)
         {
-            // update the wallets with the last processed block height
-            foreach (var wallet in this.Wallets)
-            {
-                this.UpdateLastBlockSyncedHeight(wallet, chainedBlock);
+            // update the wallets with the last processed block height                        
+            foreach(var wallet in this.Wallets) {
```
Comment (dangershony) [inline Stratis.Bitcoin/Wallet/WalletManager.cs]: You can remove the brackets all together

**Comment by dangershony** [comment]
No need lets push this fix for now and refactor later

**Comment by Neurosploit** [comment]
Ok
